### PR TITLE
use external moment parser library

### LINKED ISSUE #26: use external moment-parser
Juttle should use the external moment-parser project instead of having its own internal copy of the code.

### PR DESCRIPTION
**Pull request by demmer**
Remove the moment-parser implementation and update to use the external moment-parser library from NPM to translate moment expressions into moment or duration values.

This makes it so that Juttle no longer knows anything about the AST node types used internally within moment-parser.

Closes #26.

### CODE REVIEW

**Comment by demmer** [comment]
Oops -- there are still a couple of failing tests so this isn't quite ready yet.

**Comment by demmer** [comment]
OK this is finally ready for review. @dmajda / @bkutil please take a look.

@welch FYI in case you're interested.

File: lib/compiler/semantic.js [hunk 1]
```diff
@@ -21,6 +23,8 @@ var SemanticPass = Base.extend({
 
     initialize: function(options) {
         this.now = options.now;
+        this.beginning = moment(-Infinity);
```
Comment (welch) [inline lib/compiler/semantic.js]: et tu juttle?
per today's discussions over on moment-parser, this creates an invalid moment with NaN as its value.

File: lib/moment/juttle-moment.js [hunk 2]
```diff
@@ -844,6 +788,66 @@ var JuttleMoment = Base.extend({
...
         return operand.valueOf();
+    },
+    momentString: function(moment) {
+        if (isNaN(moment.valueOf())) {
+            return moment._i.toString();
+        } else {
+            return moment.toDate().toISOString();
+        }
+    },
+    simpleDurationString: function(d) {
```
Comment (welch) [inline lib/moment/juttle-moment.js]: around the time of shuttering, there was talk of removing this short form of duration string and only using the ASP.NET "TimeSpan" format (the thing that looks like a digital clock readout). People were often confused by this shorter form appearing in output.

Comment (dmajda) [inline lib/moment/juttle-moment.js]: This got pre-empted by open sourcing work, I hope to revisit that topic soonish.

**Comment by dmajda** [comment]
The code looks good overall. My +1 depends on the summary of grammar changes and (assuming these will be uncontroversial) related doc changes.

Also, the commits should be squashed (only 0.3.0 update, only one commit with test changes).

**Comment by welch** [comment]
It was a narrow grammar change -- eliminating "last <duration>" as a synonym of "<duration> ago", because it sounded like one was specifying a range of time rather than a single moment.

**Comment by demmer** [comment]
@welch thanks for clarifying -- I had a chat with @dmajda over on the juttle gitter room a while ago and he agreed this is a sensible thing to do.

File: lib/compiler/semantic.js (hunk 1 above)
Comment (demmer) [inline lib/compiler/semantic.js]: This is consistent with the current behavior in which we process `beginning` by calling `new JuttleMoment(-Infinity)` which in turn calls `moment(-Infinity)`.

As we discussed separately, there's a potential argument to make that we shouldn't use infinite moments for these expressions, but my goal for this PR was not to make semantic changes to the language but simply to clean up the code and reference the separate library, so I don't want to make the behavior change yet.

File: lib/moment/juttle-moment.js (hunk 2 above)
Comment (demmer) [inline lib/moment/juttle-moment.js]: That seems totally sensible to me as well -- do you know if we have an issue tracking it in the old Jira? I don't see anything in github.

Comment (dmajda) [inline lib/moment/juttle-moment.js]: PROD-8585, there is also a PR by @welch in our old repo referenced from there (could be used as a base but it has a couple of un-addressed comments).

Comment (demmer) [inline lib/moment/juttle-moment.js]: OK cool -- once we get this PR merged we can revisit that.

**Comment by demmer** [comment]
@dmajda I squashed / rebased / cleaned up as you suggested and then pushed another commit a326f16
that moves the moment stuff into the parser.

I agree that looks better - please take a look. Assuming it's ok then I would squash it with 619265b

**Comment by dmajda** [comment]
+1 with the comments addressed

**Pull request merged**